comp-bot/src/main/java/frc/robot/odometry/CustomOdometry.java

```diff
@@ -0,0 +1,183 @@
+    // positions
+    Translation2d[] moduleDisplacements = {
+      getModuleDisplacement(
+          previousWheelPositions[0].angle.getRadians(),
+          previousWheelPositions[0].distanceMeters,
+          currentWheelPositions[0].angle.getRadians(),
+          currentWheelPositions[0].distanceMeters),
+      getModuleDisplacement(
+          previousWheelPositions[1].angle.getRadians(),
+          previousWheelPositions[1].distanceMeters,
+          currentWheelPositions[1].angle.getRadians(),
+          currentWheelPositions[1].distanceMeters),
+      getModuleDisplacement(
+          previousWheelPositions[2].angle.getRadians(),
+          previousWheelPositions[2].distanceMeters,
+          currentWheelPositions[2].angle.getRadians(),
+          currentWheelPositions[2].distanceMeters),
+      getModuleDisplacement(
+          previousWheelPositions[3].angle.getRadians(),
+          previousWheelPositions[3].distanceMeters,
+          currentWheelPositions[3].angle.getRadians(),
+          currentWheelPositions[3].distanceMeters)
+    };
+
+    // Next, add the module displacements to the field relative module poses
+    Translation2d[] fieldRelativeModuleDisplacements = {
+      fieldRelativeModulePosesOfPreviousPose[0]
+          .transformBy(new Transform2d(moduleDisplacements[0], new Rotation2d(0.0)))
+          .getTranslation(),
+      fieldRelativeModulePosesOfPreviousPose[1]
+          .transformBy(new Transform2d(moduleDisplacements[1], new Rotation2d(0.0)))
+          .getTranslation(),
+      fieldRelativeModulePosesOfPreviousPose[2]
+          .transformBy(new Transform2d(moduleDisplacements[2], new Rotation2d(0.0)))
+          .getTranslation(),
+      fieldRelativeModulePosesOfPreviousPose[3]
+          .transformBy(new Transform2d(moduleDisplacements[3], new Rotation2d(0.0)))
+          .getTranslation()
+    };
+
+    // Finally, average the module displacements and return the new pose
+    Translation2d sumOfFieldRelativeModuleDisplacements =
+        fieldRelativeModuleDisplacements[0]
+            .plus(fieldRelativeModuleDisplacements[1])
+            .plus(fieldRelativeModuleDisplacements[2])
+            .plus(fieldRelativeModuleDisplacements[3]);
+    double updatedPoseX = sumOfFieldRelativeModuleDisplacements.getX() / 4.0;
+    double updatedPoseY = sumOfFieldRelativeModuleDisplacements.getY() / 4.0;
+    var updatedPose = new Pose2d(updatedPoseX, updatedPoseY, currentGyroAngle);
+
+    // Logging
+    DogLog.log("Odometry/PreviousPose", previousRobotPose);
+    DogLog.log("Odometry/PreviousWheelPositions", previousWheelPositions);
+    DogLog.log("Odometry/CurrentWheelPositions", currentWheelPositions);
+    DogLog.log("Odometry/ModuleDisplacements", moduleDisplacements);
+    DogLog.log(
+        "Odometry/FieldRelativeModuleDisplacements/FrontLeft",
+        new Pose2d(
+            fieldRelativeModuleDisplacements[0],
+            new Rotation2d(currentWheelPositions[0].angle.getRadians())));
```

Comment:
doing `new Rotation2d` with a value that is already `Rotation2d` doesn't do anything other than slow down our code